icechunk/src/change_set.rs

```diff
@@ -166,34 +276,60 @@ impl ChangeSet {
         node_id: NodeId,
         coord: ChunkIndices,
         data: Option<ChunkPayload>,
-    ) {
+    ) -> SessionResult<()> {
+        let cached_splits = self.splits.get(&node_id).expect("logic bug");
+
+        let extent = cached_splits.which_extent(&coord)?;
         // this implementation makes delete idempotent
         // it allows deleting a deleted chunk by repeatedly setting None.
         self.set_chunks
             .entry(node_id)
             .and_modify(|h| {
-                h.insert(coord.clone(), data.clone());
+                h.entry(extent.clone()).or_default().update(coord.clone(), data.clone());
             })
-            .or_insert(BTreeMap::from([(coord, data)]));
+            .or_insert_with(|| {
+                let mut h = HashMap::<ManifestExtents, SplitManifest>::with_capacity(
+                    cached_splits.len(),
+                );
+                h.entry(extent.clone())
+                    // TODO: this is duplicative. I can't use `or_default` because it's
+                    // nice to create the HashMap using `with_capacity`
+                    .or_default()
+                    .update(coord, data);
+                h
+            });
+        Ok(())
     }
 
     pub fn get_chunk_ref(
         &self,
         node_id: &NodeId,
         coords: &ChunkIndices,
     ) -> Option<&Option<ChunkPayload>> {
-        self.set_chunks.get(node_id).and_then(|h| h.get(coords))
+        self.splits
+            .get(node_id)
+            .and_then(|splits| {
+                splits.which_extent(coords).ok().map(|extent| {
+                    self.set_chunks
+                        .get(node_id)
+                        .and_then(|h| h.get(&extent))
+                        .and_then(|s| s.chunks.get(coords))
+                })
+            })
+            .flatten()
     }
 
     /// Drop the updated chunk references for the node.
     /// This will only drop the references for which `predicate` returns true
     pub fn drop_chunk_changes(
         &mut self,
         node_id: &NodeId,
-        predicate: impl Fn(&ChunkIndices) -> bool,
+        predicate: impl Fn(&ChunkIndices) -> bool + Clone,
```

Comment:
This is so weird, never had to do it .... but I can't think of a better alternative. Maybe `+ Copy` works and is more efficient?

Reply:
yeah Copy works too, I wasn't sure which to pick and why.